### DISCUSSION TITLE
nginx + skynet使用websocket的多吗

### DISCUSSION THREAD

**Discussion by JavenLaw** [cloudwu/skynet#1994] [Ideas]
最近想搞一下skynet中websocket，skynet自带的simplewebsocket.lua可以直接使用，但具体怎么引入工程变得好用还没头绪

看了 hanxi 的 [websocket-gate reponse转发模式](https://github.com/hanxi/skynet-demo/blob/master/service/ws_gate.lua)

特点：采用了转发的模式推送数据给客户端，为了解决热点问题，gate采用master/slave

初步的想法：
还是每个agent自己有websocket.lua，跟[ websocket-gate非转发模式](https://github.com/xzhovo/skynet-websocket-gate) 一样
但是每个agent和nginx连接，用ws就行
nginx和client连接用wss

好处是：
1，nginx可以负载均衡，方便分配到不同的机器上去
2，skynet不用考虑https，wss证书的问题
3，nginx负责类似gate的功能，自己不用维护gate和client的网络连接了
（但是不知道 nginx<--->agent的连接管理 和  gate<--->client的连接管理 会不会差不多？如果差不多，这点就没意义了，哈哈）

不知道这样的想法和设计有什么大的问题
或者大家在实践中有其他更好的方案？
希望能讨论一下~

**Comment by hanxi** [comment]
可以考虑用 goscon 做游戏服的网关，这里有个支持 ws 的 pr ： https://github.com/ejoy/goscon/pull/38/files

**Reply by JavenLaw** [reply]
好，我先了解一下

**Comment by changlongH** [comment]
nginx<--->agent 没问题的，稳定且高效。同时NG也支持多种负载均衡策略和热更新。

**Reply by JavenLaw** [reply]
嗯嗯
不过当client <---> nginx的连接断开/重连，nginx<--->agent的连接状态是能自动更新吗？
换句话说，client <---> nginx  和 nginx<--->agent 连接是不是分别独立的？

**Reply by changlongH** [reply]
是分别独立，你可以把NG当成透明的代理。只需要处理好client ip 的问题。

**Comment by JieTrancender** [comment]
SLB -> openresty -> gate。我们当初这么做是因为要wss，然后gate节点多，用openresty写脚本方便更新gate节点信息。

**Reply by JavenLaw** [reply]
有一个疑问，流量不大的话
有了openresty，是不是可以不用gate
直接SLB -> openresty -> agent_1
--------------------------> agent_2
--------------------------> agent_3
实际openresty就是功能增强的nginx

**Reply by hanxi** [reply]
用 openresty 的话，skynet 直接用 tcp 那个 gate 都行，不用限制 ws 了。

**Reply by JavenLaw** [reply]
> 用 openresty 的话，skynet 直接用 tcp 那个 gate 都行，不用限制 ws 了。

是哪个呢？可以指一下文件吗:）

**Reply by hanxi** [reply]
就是示例里的 tcp 链接的这个 https://github.com/cloudwu/skynet/blob/master/examples/watchdog.lua 。

client --ws--> openresty --tcp--> gate

**Reply by JavenLaw** [reply]
> 就是示例里的 tcp 链接的这个 https://github.com/cloudwu/skynet/blob/master/examples/watchdog.lua 。
> 
> client --ws--> openresty --tcp--> gate

感谢！
